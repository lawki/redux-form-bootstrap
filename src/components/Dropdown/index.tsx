import * as React from 'react';
import { ButtonDropdown, DropdownToggle, DropdownMenu, DropdownItem  } from 'reactstrap';
import { BaseFieldProps } from '../../common/interfaces/CommonProps';

export interface DropdownOption{
    value: string, 
    type: 'header' | 'disabled' | 'divider' 
}
export interface DropdownProps extends BaseFieldProps{
    defaultValue?: any;
    input: any;
    isCaret?: boolean;
    color?: 'primary' | 'secondary' | 'success' | 'info' | 'warning' | 'danger',
    text: string;
    options: DropdownOption[];
    size: 'lg' | 'sm',
    direction: 'up' | 'down' | 'left' | 'right'
}

export interface RFDropdownState {
    isOpen: boolean;
}

export class Dropdown extends React.Component<DropdownProps, RFDropdownState>{

    static defaultProps = {
        isCaret: true,
        color: 'info',
        size: 'sm',
        direction: 'down'
    }

    static DROPDOWN_TYPE = {
        HEADER: 'header',
        DISABLED: 'disabled',
        DIVIDER: 'divider',
    }

    constructor(props: DropdownProps){
        super(props);
        this.state = {
            isOpen: false
        }
    }

    toggle = ()=> this.setState((oldState: { isOpen: any; })=> ({isOpen: !oldState.isOpen}))

    saveToStore = ( data : string)=>{
        const { filter, input } = this.props;
        data = filter && typeof filter === typeof Function ? filter(data): data;
        input.onChange(data);
    }

    render(){
        const { isCaret, color, options, text, size, direction } = this.props;
        const { isOpen } = this.state;
        return (
            <ButtonDropdown isOpen={isOpen} toggle={this.toggle} direction={direction}>
            <DropdownToggle caret={isCaret} color={color} size={size}>
              {text}
            </DropdownToggle>
            <DropdownMenu>
                {options && Array.isArray(options) ? 
                options.map( option => {
                    const types = Dropdown.DROPDOWN_TYPE;
                    const handler = ()=> this.saveToStore(option.value);
                    switch(option.type){
                        case types.DISABLED:
                            return <DropdownItem onClick={handler} disabled >{option.value}</DropdownItem>;
                        case types.DIVIDER:
                            return <DropdownItem  onClick={handler} divider>{option.value}</DropdownItem>;
                        case types.HEADER:
                                return <DropdownItem  onClick={handler} header>{option.value}</DropdownItem>;
                        default:
                            return <DropdownItem  onClick={handler}>{option.value}</DropdownItem>;
                            
                    }
                }): null}
            </DropdownMenu>
          </ButtonDropdown>
        )
    }
}