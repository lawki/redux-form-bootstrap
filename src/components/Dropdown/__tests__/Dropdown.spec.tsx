import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Dropdown } from '..';
import { options } from './Dropdown.options';
describe('Dropdown Component', () => {

    it('renders', () => {
        const fn = jest.fn();
        const app = shallow(
            <Dropdown
                text="Select a name"
                options={options}
                onChange={fn}
            />);
        expect(app).toMatchSnapshot();
    });

    it('has a single .btn-group', () => {
        const fn = jest.fn();
        const app = mount(
            <Dropdown
                text="Select a name"
                options={options}
                onChange={fn}
            />);
        expect(app.find('.btn-group')).toHaveLength(1);
    });
});