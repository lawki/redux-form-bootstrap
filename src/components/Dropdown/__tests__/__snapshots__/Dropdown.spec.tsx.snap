// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component renders 1`] = `
<ButtonDropdown
  direction="down"
  isOpen={false}
  toggle={[Function]}
>
  <DropdownToggle
    aria-haspopup={true}
    caret={true}
    color="info"
    size="sm"
  >
    Select a name
  </DropdownToggle>
  <DropdownMenu
    flip={true}
    tag="div"
  >
    <DropdownItem
      disabled={true}
      key="You haven't selected a name yet!"
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
      You haven't selected a name yet!
    </DropdownItem>
    <DropdownItem
      header={true}
      key="Karen"
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
      Karen
    </DropdownItem>
    <DropdownItem
      divider={true}
      onClick={[Function]}
      tag="button"
      toggle={true}
    />
    <DropdownItem
      header={true}
      key="Gillian"
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
      Gillian
    </DropdownItem>
    <DropdownItem
      divider={true}
      onClick={[Function]}
      tag="button"
      toggle={true}
    />
    <DropdownItem
      header={true}
      key="Bucky"
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
      Bucky
    </DropdownItem>
    <DropdownItem
      divider={true}
      onClick={[Function]}
      tag="button"
      toggle={true}
    />
    <DropdownItem
      header={true}
      key="Bikrum"
      onClick={[Function]}
      tag="button"
      toggle={true}
    >
      Bikrum
    </DropdownItem>
    <DropdownItem
      divider={true}
      onClick={[Function]}
      tag="button"
      toggle={true}
    />
  </DropdownMenu>
</ButtonDropdown>
`;
